src/python/gudhi/_pers_cub_low_dim.cc

```diff
@@ -5,84 +5,109 @@
  *    Copyright (C) 2023 Inria
  *
  *    Modification(s):
+ *      - 2025/01 Vincent Rouvreau: Use nanobind instead of PyBind11 for python bindings
  *      - YYYY/MM Author: Description of the modification
  */
 
+// #include <array>
+#include <cstddef>
 #include <vector>
-#include <array>
 #include <limits>
-#include <stdexcept>
 
-#include <pybind11/pybind11.h>
-#include <pybind11/stl.h>
-#include <pybind11/stl_bind.h>
-#include <pybind11/numpy.h>
+#include <nanobind/nanobind.h>
+#include <nanobind/ndarray.h>
 
 #include <boost/range/counting_range.hpp>
 #include <boost/range/adaptor/transformed.hpp>
 
 #include <gudhi/Persistence_on_a_line.h>
 #include <gudhi/Persistence_on_rectangle.h>
-#include <gudhi/Debug_utils.h>
+#include <python_interfaces/numpy_utils.h>
 
-namespace py = pybind11;
-typedef std::vector<std::array<float, 2>> Vf;
-typedef std::vector<std::array<double, 2>> Vd;
-PYBIND11_MAKE_OPAQUE(Vf);
-PYBIND11_MAKE_OPAQUE(Vd);
+namespace nb = nanobind;
 
 template <class T>
-py::array wrap_persistence_1d(py::array_t<T> data)
+auto wrap_persistence_1d(nb::ndarray<T, nb::ndim<1>> data)
 {
-  py::buffer_info buf = data.request();
-  if (buf.ndim != 1) throw std::runtime_error("Data must be a 1-dimensional array");
-  auto cnt = boost::counting_range<py::ssize_t>(0, buf.shape[0]);
-  char* p = static_cast<char*>(buf.ptr);
-  auto h = buf.strides[0];
-  auto proj = [=](py::ssize_t i) { return *reinterpret_cast<T*>(p + i * h); };
+  auto data_view = data.view();
+  auto cnt = boost::counting_range<nb::ssize_t>(0, data_view.shape(0));
+  auto proj = [&](nb::ssize_t i) { return data_view(i); };
   auto r = boost::adaptors::transform(cnt, proj);
-  std::vector<std::array<T, 2>> dgm;
+  auto dgm = new T[data_view.shape(0) * 2];
```

Comment:
Isn't 2*n a bit much? I'd have to check, but I think n+1 should suffice, each input element is output at most once, and +1 for infinity that is paired with the global minimum.

Reply:
I just had in mind the worst case where nothing dies, so every simplex is paired with infinity. But in this particular case, I guess it never happens? Because you always have a full "surface of cubes/squares", like for `Bitmap_cubical_complex` ?

Reply:
In the end, you have a segment. That's contractible, so just 1 for Betti₀ and 0 for the rest.